File: Sep/Day6/minimum-operations-to-make-array-elements-zero.py
#  Time: O(N * Log(R))  :: Space: O(1)
class Solution:
    def minimumOperations(self, num: int) -> int:
        binaryLength = 1
        base = 1
        count = 0
        while base <= num:
            count += ((binaryLength + 1) // 2) * (min(base * 2 - 1, num) - base + 1)
            binaryLength += 1
            base *= 2
        return count

    def minOperations(self, queries: List[List[int]]) -> int:
        sumOfResults = 0
        for query in queries:
            sumOfResults += (self.minimumOperations(query[1]) - self.minimumOperations(query[0] - 1) + 1) // 2
        return sumOfResults

File: Sep/Day6/minimum-operations-to-make-array-elements-zero.cc
//  Time: O(N * Log(R))  :: Space: O(1)
class Solution {
public:
    long long minimumOperations(long long num) {
        long long binaryLength = 1;
        long long base = 1;
        long long count = 0;
        while (base <= num) {
            count += ((binaryLength + 1) / 2) * (min(base * 2 - 1, num) - base + 1);
            binaryLength++;
            base *= 2;
        }
        return count;
    }

    long long minOperations(vector<vector<long long>>& queries) {
        long long sumOfResults = 0;
        for (auto& query : queries) {
            sumOfResults += (minimumOperations(query[1]) - minimumOperations(query[0] - 1) + 1) / 2;
        }
        return sumOfResults;
    }
};
